File: java-sdk/radar-schemas-registration/src/main/java/org/radarbase/schema/registration/KafkaTopics.java
package org.radarbase.schema.registration;

import static org.apache.kafka.clients.admin.AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG;

import java.io.IOException;
import java.io.InputStream;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Collection;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Properties;
import java.util.Set;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.regex.Pattern;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import javax.validation.constraints.NotNull;
import org.apache.kafka.clients.admin.AdminClient;
import org.apache.kafka.clients.admin.ListTopicsOptions;
import org.apache.kafka.clients.admin.NewTopic;
import org.apache.kafka.common.Node;
import org.radarbase.schema.specification.SourceCatalogue;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Registers Kafka topics with Zookeeper.
 */
public class KafkaTopics implements TopicRegistrar {

    private static final Logger logger = LoggerFactory.getLogger(KafkaTopics.class);
    private final Map<String, Object> kafkaProperties;
    private final AdminClient kafkaClient;
    private boolean initialized;
    static final int MAX_SLEEP = 32;
    private Set<String> topics;

    /**
     * Create Kafka topics registration object with given Zookeeper.
     *
     * @param kafkaProperties Kafka properties that must include a bootstrap.servers
     *                        or zookeeper.connect property.
     */
    public KafkaTopics(@NotNull Map<String, Object> kafkaProperties) {
        initialized = false;
        this.kafkaProperties = kafkaProperties;
        this.kafkaClient = AdminClient.create(kafkaProperties);
    }

    /**
     * Wait for brokers to become available. This uses a polling mechanism, waiting for at most 200
     * seconds.
     *
     * @param brokers number of brokers to wait for
     * @throws InterruptedException when waiting for the brokers is interrupted.
     */
    @Override
    public void initialize(int brokers) throws InterruptedException {
        initialize(brokers, 20);
    }

    /**
     * Wait for brokers to become available. This uses a polling mechanism, retrying with sleep
     * up to the supplied numTries on failures. The sleep time is doubled every retry
     * iteration until the {@value #MAX_SLEEP} is reached which then takes precedence.
     *
     * @param brokers number of brokers to wait for.
     * @param numTries Number of times to retry in case of failure.
     * @throws InterruptedException when waiting for the brokers is interrupted.
     */
    @Override
    public void initialize(int brokers, int numTries) throws InterruptedException {
        int sleep = 2;
        int numBrokers = 0;

        for (int tries = 0; tries < numTries && numBrokers < brokers; tries++) {
            List<String> hosts;
            try {
                hosts = kafkaClient.describeCluster()
                        .nodes()
                        .get()
                        .stream()
                        .map(Node::host)
                        .collect(Collectors.toList());
            } catch (ExecutionException ex) {
                logger.error("Failed to connect to bootstrap server {}",
                        kafkaProperties.get(BOOTSTRAP_SERVERS_CONFIG), ex.getCause());
                hosts = Collections.emptyList();
            }
            numBrokers = hosts.size();

            if (numBrokers >= brokers) {
                break;
            } else if (tries < numTries - 1) {
                logger.warn("Only {} out of {} Kafka brokers available. Waiting {} seconds.",
                        numBrokers, brokers, sleep);
                Thread.sleep(sleep * 1000L);
                sleep = Math.min(MAX_SLEEP, sleep * 2);
            } else {
                logger.error("Only {} out of {} Kafka brokers available."
                        + " Failed to wait on all brokers.", numBrokers, brokers);
            }
        }

        initialized = numBrokers >= brokers;

        if (!initialized || !refreshTopics()) {
            throw new IllegalStateException("Brokers or topics not available.");
        }
    }

    @Override
    public void ensureInitialized() {
        if (!initialized) {
            throw new IllegalStateException("Manager is not initialized yet");
        }
    }

    @Override
    public int createTopics(@NotNull SourceCatalogue catalogue, int partitions, short replication,
            String topic, String match) {
        Pattern pattern = TopicRegistrar.matchTopic(topic, match);

        if (pattern == null) {
            return createTopics(catalogue, partitions, replication) ? 0 : 1;
        } else {
            List<String> topicNames =
                    catalogue.getTopicNames().filter(s -> pattern.matcher(s).find())
                            .collect(Collectors.toList());

            if (topicNames.isEmpty()) {
                logger.error("Topic {} does not match a known topic."
                        + " Find the list of acceptable topics"
                        + " with the `radar-schemas-tools list` command. Aborting.", pattern);
                return 1;
            }
            return createTopics(topicNames.stream(), partitions, replication) ? 0 : 1;
        }
    }

    /**
     * Create all topics in a catalogue.
     *
     * @param catalogue source catalogue to extract topic names from
     * @param partitions number of partitions per topic
     * @param replication number of replicas for a topic
     * @return whether the whole catalogue was registered
     */
    private boolean createTopics(@NotNull SourceCatalogue catalogue, int partitions,
            short replication) {
        ensureInitialized();
        return createTopics(catalogue.getTopicNames(), partitions, replication);
    }

    @Override
    public boolean createTopics(Stream<String> topicsToCreate, int partitions, short replication) {
        ensureInitialized();
        try {
            refreshTopics();
            logger.info("Creating topics. Topics marked with [*] already exist.");

            List<NewTopic> newTopics = topicsToCreate.sorted().distinct().filter(t -> {
                if (this.topics != null && this.topics.contains(t)) {
                    logger.info("[*] {}", t);
                    return false;
                } else {
                    logger.info("[ ] {}", t);
                    return true;
                }
            }).map(t -> new NewTopic(t, partitions, replication)).collect(Collectors.toList());

            if (!newTopics.isEmpty()) {
                getKafkaClient().createTopics(newTopics).all().get();
                logger.info("Created {} topics. Requesting to refresh topics", newTopics.size());
                refreshTopics();
            } else {
                logger.info("All of the topics are already created.");
            }
            return true;
        } catch (Exception ex) {
            logger.error("Failed to create topics {}", ex.toString());
            return false;
        }
    }

    @Override
    public boolean refreshTopics() throws InterruptedException {
        ensureInitialized();
        logger.info("Waiting for topics to become available.");
        int sleep = 10;
        int numTries = 10;

        topics = null;
        ListTopicsOptions opts = new ListTopicsOptions().listInternal(true);
        for (int tries = 0; tries < numTries; tries++) {
            try {
                topics = getKafkaClient().listTopics(opts).names().get(sleep, TimeUnit.SECONDS);
            } catch (ExecutionException e) {
                logger.error("Failed to list topics from brokers: {}."
                        + " Trying again after {} seconds.", e, sleep);
                Thread.sleep(sleep * 1000L);
                sleep = Math.min(MAX_SLEEP, sleep * 2);
                continue;
            } catch (TimeoutException e) {
                // do nothing
            }
            if (topics != null && !topics.isEmpty()) {
                break;
            }
            if (tries < numTries - 1) {
                logger.warn("Topics not listed yet after {} seconds", sleep);
            } else {
                logger.error("Topics have not become available. Failed to wait on Kafka.");
            }
            sleep = Math.min(MAX_SLEEP, sleep * 2);
        }

        if (topics == null || topics.isEmpty()) {
            return false;
        } else {
            Thread.sleep(5000L);
            return true;
        }
    }

    @Override
    public Set<String> getTopics() {
        ensureInitialized();
        return Collections.unmodifiableSet(topics);
    }


    @Override
    public void close() {
        kafkaClient.close();
    }

    /**
     * Get current number of Kafka brokers according to Zookeeper.
     *
     * @return number of Kafka brokers
     * @throws ExecutionException if kafka cannot connect
     * @throws InterruptedException if the query is interrupted.
     */
    public int getNumberOfBrokers() throws ExecutionException, InterruptedException {
        return kafkaClient.describeCluster()
                .nodes()
                .thenApply(Collection::size)
                .get();
    }

    @NotNull
    @Override
    public AdminClient getKafkaClient() {
        ensureInitialized();
        return kafkaClient;
    }

    @Override
    public Map<String, Object> getKafkaProperties() {
        return kafkaProperties;
    }

    public static Map<String, Object> loadConfig(
            final String configFile,
            String bootstrapServers
    ) throws IOException {
        Map<String, Object> kafkaConfig;
        if (configFile != null && !configFile.isEmpty()) {
            final Properties cfg = new Properties();
            try (InputStream inputStream = Files.newInputStream(Paths.get(configFile))) {
                cfg.load(inputStream);
            }
            kafkaConfig = cfg.entrySet().stream()
                    .collect(Collectors.toMap(e -> (String)e.getKey(), Entry::getValue));
        } else {
            kafkaConfig = new HashMap<>();
        }
        if (bootstrapServers != null && !bootstrapServers.isEmpty()) {
            kafkaConfig.put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        }
        if (kafkaConfig.get(BOOTSTRAP_SERVERS_CONFIG) == null) {
            throw new IllegalStateException("Cannot configure Kafka without "
                    + BOOTSTRAP_SERVERS_CONFIG + " property");
        }
        return kafkaConfig;
    }
}

File: java-sdk/radar-schemas-registration/src/main/java/org/radarbase/schema/registration/KafkaTopics.kt
package org.radarbase.schema.registration

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.withContext
import org.apache.kafka.clients.admin.Admin
import org.apache.kafka.clients.admin.AdminClient
import org.apache.kafka.clients.admin.AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG
import org.apache.kafka.clients.admin.ListTopicsOptions
import org.apache.kafka.clients.admin.NewTopic
import org.apache.kafka.common.config.SaslConfigs.SASL_JAAS_CONFIG
import org.radarbase.kotlin.coroutines.suspendGet
import org.radarbase.schema.specification.SourceCatalogue
import org.radarbase.schema.specification.config.ToolConfig
import org.radarbase.schema.specification.config.TopicConfig
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeoutException
import java.util.stream.Collectors
import java.util.stream.Stream
import kotlin.time.Duration.Companion.seconds
import kotlin.time.TimeMark
import kotlin.time.TimeSource

/**
 * Registers Kafka topics with Zookeeper.
 */
class KafkaTopics(
    private val toolConfig: ToolConfig,
) : TopicRegistrar {
    private var initialized = false
    override lateinit var topics: Set<String>

    private val adminClient: AdminClient = AdminClient.create(toolConfig.kafka)

    /**
     * Wait for brokers to become available. This uses a polling mechanism, waiting for at most 200
     * seconds.
     *
     * @param brokers number of brokers to wait for
     * @throws InterruptedException when waiting for the brokers is interrupted.
     */
    override suspend fun initialize(brokers: Int) {
        initialize(brokers, 20)
    }

    /**
     * Wait for brokers to become available. This uses a polling mechanism, retrying with sleep
     * up to the supplied numTries on failures. The sleep time is doubled every retry
     * iteration until the {@value #MAX_SLEEP} is reached which then takes precedence.
     *
     * @param brokers number of brokers to wait for.
     * @param numTries Number of times to retry in case of failure.
     */
    override suspend fun initialize(brokers: Int, numTries: Int) {
        val numBrokers = retryFlow(2.seconds, MAX_SLEEP)
            .take(numTries)
            .map { sleep ->
                try {
                    withContext(Dispatchers.IO) {
                        adminClient.describeCluster()
                            .nodes()
                            .suspendGet(sleep)
                            .size
                    }
                } catch (ex: InterruptedException) {
                    logger.error("Refreshing topics interrupted")
                    throw ex
                } catch (ex: TimeoutException) {
                    logger.error(
                        "Failed to connect to bootstrap server {} within {} seconds",
                        kafkaProperties[BOOTSTRAP_SERVERS_CONFIG],
                        sleep,
                    )
                    0
                } catch (ex: Throwable) {
                    logger.error(
                        "Failed to connect to bootstrap server {}",
                        kafkaProperties[BOOTSTRAP_SERVERS_CONFIG],
                        ex.cause,
                    )
                    0
                }
            }
            .firstOrNull { numBrokers ->
                if (numBrokers >= brokers) {
                    true
                } else {
                    logger.warn("Only {} out of {} Kafka brokers available.", numBrokers, brokers)
                    false
                }
            }

        initialized = numBrokers != null
        check(initialized) { "Brokers not available." }
        check(refreshTopics()) { "Topics not available." }
    }

    override fun ensureInitialized() {
        check(initialized) { "Manager is not initialized yet" }
    }

    override suspend fun createTopics(
        catalogue: SourceCatalogue,
        partitions: Int,
        replication: Short,
        topic: String?,
        match: String?,
    ): Int {
        val pattern = TopicRegistrar.matchTopic(topic, match)
        return if (pattern == null) {
            if (createTopics(catalogue, partitions, replication)) 0 else 1
        } else {
            val topicNames = topicNames(catalogue)
                .filter { s -> pattern.matcher(s).find() }
                .collect(Collectors.toList())
            if (topicNames.isEmpty()) {
                logger.error(
                    "Topic {} does not match a known topic." +
                        " Find the list of acceptable topics" +
                        " with the `radar-schemas-tools list` command. Aborting.",
                    pattern,
                )
                return 1
            }
            if (createTopics(topicNames.stream(), partitions, replication)) 0 else 1
        }
    }

    private fun topicNames(catalogue: SourceCatalogue): Stream<String> {
        return Stream.concat(
            catalogue.topicNames,
            toolConfig.topics.keys.stream(),
        ).filter { t -> toolConfig.topics[t]?.enabled != false }
    }

    /**
     * Create all topics in a catalogue.
     *
     * @param catalogue source catalogue to extract topic names from
     * @param partitions number of partitions per topic
     * @param replication number of replicas for a topic
     * @return whether the whole catalogue was registered
     */
    private suspend fun createTopics(
        catalogue: SourceCatalogue,
        partitions: Int,
        replication: Short,
    ): Boolean {
        ensureInitialized()
        return createTopics(topicNames(catalogue), partitions, replication)
    }

    override suspend fun createTopics(
        topics: Stream<String>,
        partitions: Int,
        replication: Short,
    ): Boolean {
        ensureInitialized()
        return try {
            refreshTopics()
            logger.info("Creating topics. Topics marked with [*] already exist.")
            val newTopics = topics
                .sorted()
                .distinct()
                .filter { t: String ->
                    if (this.topics.contains(t)) {
                        logger.info("[*] {}", t)
                        return@filter false
                    } else {
                        logger.info("[ ] {}", t)
                        return@filter true
                    }
                }
                .map { t ->
                    val topicConfig = toolConfig.topics[t]
                        ?: TopicConfig()
                    NewTopic(
                        t,
                        topicConfig.partitions ?: partitions,
                        topicConfig.replicationFactor ?: replication,
                    ).configs(topicConfig.properties)
                }
                .collect(Collectors.toList())

            if (newTopics.isNotEmpty()) {
                kafkaClient
                    .createTopics(newTopics)
                    .all()
                    .suspendGet()
                logger.info("Created {} topics. Requesting to refresh topics", newTopics.size)
                refreshTopics()
            } else {
                logger.info("All of the topics are already created.")
            }
            true
        } catch (ex: Exception) {
            logger.error("Failed to create topics {}", ex.toString())
            false
        }
    }

    @Throws(InterruptedException::class)
    override suspend fun refreshTopics(): Boolean {
        ensureInitialized()
        logger.info("Waiting for topics to become available.")

        topics = emptySet()
        val opts = ListTopicsOptions().apply {
            listInternal(true)
        }

        topics = retryFlow(2.seconds, MAX_SLEEP)
            .take(10)
            .map { sleep ->
                try {
                    kafkaClient
                        .listTopics(opts)
                        .names()
                        .suspendGet(sleep)
                } catch (ex: TimeoutException) {
                    logger.error("Failed to list topics within {} seconds", sleep)
                    emptySet()
                } catch (ex: InterruptedException) {
                    logger.error("Refreshing topics interrupted")
                    throw ex
                } catch (ex: Throwable) {
                    logger.error("Failed to list topics from brokers: {}", ex.cause.toString())
                    emptySet()
                }
            }
            .firstOrNull { it.isNotEmpty() }
            ?: emptySet()

        return topics.isNotEmpty()
    }

    override fun close() {
        adminClient.close()
    }

    /**
     * Get current number of Kafka brokers according to Zookeeper.
     *
     * @return number of Kafka brokers
     */
    suspend fun numberOfBrokers(): Int {
        return adminClient.describeCluster()
            .nodes()
            .suspendGet()
            .size
    }

    override val kafkaClient: Admin
        get() {
            ensureInitialized()
            return adminClient
        }

    override val kafkaProperties: Map<String, *>
        get() = toolConfig.kafka

    companion object {
        private val logger = LoggerFactory.getLogger(KafkaTopics::class.java)
        private val MAX_SLEEP = 32.seconds

        @JvmStatic
        fun ToolConfig.configureKafka(
            bootstrapServers: String?,
        ): ToolConfig = if (bootstrapServers.isNullOrEmpty()) {
            check(BOOTSTRAP_SERVERS_CONFIG in kafka) {
                "Cannot configure Kafka without $BOOTSTRAP_SERVERS_CONFIG property"
            }
            this
        } else {
            copy(
                kafka = buildMap {
                    putAll(kafka)
                    put(BOOTSTRAP_SERVERS_CONFIG, bootstrapServers)
                    System.getenv("KAFKA_SASL_JAAS_CONFIG")?.let {
                        put(SASL_JAAS_CONFIG, it)
                    }
                },
            )
        }

        fun retryFlow(
            startSleep: kotlin.time.Duration,
            maxSleep: kotlin.time.Duration,
        ): Flow<kotlin.time.Duration> = flow {
            var sleep = startSleep

            while (true) {
                // All computation for the sequence will be done in yield. It should be excluded
                // from sleep.
                val endTime = TimeSource.Monotonic.markNow() + sleep
                emit(sleep)
                sleepUntil(endTime) { timeUntil ->
                    logger.info("Waiting {} seconds to retry", timeUntil)
                }
                if (sleep < maxSleep) {
                    sleep = (sleep * 2).coerceAtMost(maxSleep)
                }
            }
        }

        private suspend fun sleepUntil(time: TimeMark, beforeSleep: (kotlin.time.Duration) -> Unit) {
            val timeUntil = -time.elapsedNow()
            if (timeUntil.isPositive()) {
                beforeSleep(timeUntil)
                delay(timeUntil)
            }
        }
    }
}
